### DISCUSSION TITLE
How to Create a Subdag Using Taskflow API

### DISCUSSION THREAD

**Discussion by RaedShabbir** [apache/airflow#15403] [Q&A]
Hello, 

I would like to construct a subdag using the taskflow api but am a bit confused of how to go about this. Currently I have two dags, the first loads N number of items to be processed. Each item would then be passed into its own subdag for parallel processing. The second dag (subdag) would take care of all processing steps for an individual item, with number of subdags == number of processors.   

What would be the best way to accomplish this? Would I be able to create a subdag by just returning the functional dag? 

e.g. in the same directory as the subdag
```
@dag (....
)
def subdag_1(parameter_1):
    ......

def create_subdag(parameter_1):
    return subdag_1(parameter_1)
```

Then use the above function to create a SubDagOperator?

Or am I going about parallelism the wrong way? 

Regards, 
Raed

**Comment by RaedShabbir** [comment]
@turbaszek Any idea on this?

**Reply by turbaszek** [reply]
Do you need subdag? Why not task group?

**Reply by RaedShabbir** [reply]
You're recommending instead to keep a single dag and turn all of the tasks that need to be run in parallel into a task group?

But how would we define a taskgroup and order of tasks inside the TaskflowAPI? 

Would the code here work fine from within the taskflow API? 
https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#taskgroup

Or would we need to modify with something similar? 
```py
@task
def function_1(param):
   pass 


@task
def function_2(param):
   pass 

with TaskGroup("group1") as group1:
    output_1 = function_1
    output_2  = function_2(output_1)

group_1
```

Lastly how would we actually get the parallel operations? Would the following be enough for automatic parallelism? 
With this running as many taskgroups in parallel as specified by airflow.cfg? 

for item in items: 
   group_1(item)

**Reply by turbaszek** [reply]
The TaskFlow is fully working with TaskGroups as far as I know, we even have #15034 in progress that will improve the experience.

> Lastly how would we actually get the parallel operations? Would the following be enough for automatic parallelism?

If the task group is supposed to be configurable I would suggest doing something like
```py
def tg_factory(input_value):
  with TaskGroup(..) as tg:
    # do some logic
  return tg


with DAG(...) as dag:
  for input_value in many:
    tg_factory(input_value)
```
